src/dynamixel_hardware_interface.cpp

```diff
@@ -144,24 +179,41 @@ namespace dynamixel
     **/
     void DynamixelHardwareInterface::write_joints()
     {
-        std::vector<int> command_int;
-        command_int.resize(_dynamixel_ids.size());
 
-        for(unsigned int i=0; i<_dynamixel_ids.size(); i++)
-        {
-            command_int[i] = (int)(_joint_commands[i]*2048/M_PI) + 2048;
+        for (unsigned int i = 0; i < _dynamixel_servos.size(); i++) {
+            std::map<long long int, std::string>::iterator dynamixel_iterator = _dynamixel_map.find(_dynamixel_servos[i]->id());
+            if (dynamixel_iterator == _dynamixel_map.end())
+                continue;
+            // Sending commands only when needed
+            if (std::abs(_joint_commands[i] - _prev_commands[i]) < std::numeric_limits<double>::epsilon())
+                continue;
+            _prev_commands[i] = _joint_commands[i];
+            try {
+                dynamixel::StatusPacket<dynamixel::protocols::Protocol1> status;
+
+                double goal_pos = _joint_commands[i];
+
+                std::map<long long int, double>::iterator dynamixel_corrections_iterator = _dynamixel_corrections.find(_dynamixel_servos[i]->id());
+                if (dynamixel_corrections_iterator != _dynamixel_corrections.end()) {
+                    goal_pos += dynamixel_corrections_iterator->second;
+                }
+
+                ROS_DEBUG_STREAM("Setting position of joint " << _dynamixel_servos[i]->id() << " to " << goal_pos);
+                _dynamixel_controller.send(_dynamixel_servos[i]->reg_goal_position_angle(goal_pos));
+                _dynamixel_controller.recv(status);
+            }
+            catch (dynamixel::errors::Error& e) {
+                ROS_ERROR_STREAM("Caught a Dynamixel exception while sending new commands:\n"
+                    << e.msg());
+            }
         }
 
-        try
-        {
-            dynamixel::Status status;
-            _dynamixel_controller.send(dynamixel::ax12::SetPositions(_dynamixel_ids, command_int));
-            _dynamixel_controller.recv(_read_timeout, status);
+        try {
+            _dynamixel_controller.send(dynamixel::instructions::Action<dynamixel::protocols::Protocol1>(dynamixel::protocols::Protocol1::broadcast_id));
```

Comment:
OK, I'll have a look at this. I'll need to document this otherwise we'll never know when to use recv and when not.